test_heterogeneous_sequences.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/env python3
+
+from unittest import TestCase
+from dark.reads import Read, Reads
+from collections import Counter, defaultdict
+from heterogeneous_sequences import heterogeneousSites, compareToRef
+
+
+class TestHeterogeneousSites(TestCase):
+    """
+    Test the heterogeneousSites function.
+    """
+
+    def testOneRead(self):
+        read = Read('id', 'ACCG')
+        reads = Reads([read])
+
+        self.assertEqual({}, heterogeneousSites(reads, len(read), 1))
+
+    def testHomogeneousReads(self):
+        read = Read('id', 'ACCG')
+        reads = Reads([read, Read('id2', 'ACCG')])
+
+        self.assertEqual({}, heterogeneousSites(reads, len(read), 1))
+
+    def testHeterogeneousReadsOneDifference(self):
+        read = Read('id', 'ACCG')
+        reads = Reads([read, Read('id2', 'ACCC')])
+
+        self.assertEqual({3: Counter({'G': 1, 'C': 1})}, heterogeneousSites(reads, len(read), 1))
+
+    def testHeterogeneousReadsTwoDifferences(self):
+        read = Read('id', 'ACCG')
+        reads = Reads([read, Read('id2', 'TCCC')])
+
+        self.assertEqual({0: Counter({'A': 1, 'T': 1}), 3: Counter({'G': 1, 'C': 1})}, heterogeneousSites(reads, len(read), 1))
+
+    def testHeterogeneousReadsFractionHigh(self):
+        read = Read('id', 'ACCG')
+        reads = Reads([read, Read('id2', 'ACCC'), Read('id3', 'ACCC')])
+
+        self.assertEqual({3: Counter({'G': 1, 'C': 2})}, heterogeneousSites(reads, len(read), 0.7))
+
+    def testHeterogeneousReadsFractionLow(self):
+        read = Read('id', 'ACCG')
+        reads = Reads([read, Read('id2', 'ACCC'), Read('id3', 'ACCC')])
+
+        self.assertEqual({}, heterogeneousSites(reads, len(read), 0.6))
+
+    def testHeterogeneousReadsFractionLowWithOneDifference(self):
+        read = Read('id', 'ACCG')
+        reads = Reads([read, Read('id2', 'TCCG'), Read('id3', 'TCCG'), Read('id4', 'ACCG')])
+
+        self.assertEqual({0: Counter({'A': 2, 'T': 2})}, heterogeneousSites(reads, len(read), 0.6))
+
+    def testGaps(self):
+        read = Read('id', 'ACCG')
+        reads = Reads([read, Read('id2', '-CC-')])
+
+        self.assertEqual({}, heterogeneousSites(reads, len(read), 1))
+
+    def testEmptyReads(self):
```

Comment:
Can you put in a comment explaining why this would raise a `NameError`?  I don't get it.

Reply:
That was a mistake - len(reads) didn't refer to anything, as there was no read, and so the NameError was because read was not defined. In this case, instead an empty dictionary is expected. (Changed this)